File: labs/12-lsp-treesitter/java/Main.java
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Main {

    private static final String SAMPLE_CODE = """
            \"\"\"A tiny module for demonstration.\"\"\"

            import json
            from datetime import datetime


            def greet(name: str) -> str:
                \"\"\"Return a friendly greeting.\"\"\"
                return f"Hello, {name}!"


            class Calculator:
                \"\"\"Simple calculator.\"\"\"

                def add(self, a: float, b: float) -> float:
                    return a + b

                def subtract(self, a: float, b: float) -> float:
                    return a - b


            def main():
                calc = Calculator()
                print(greet("world"))
                print(calc.add(1, 2))
            """;

    private static final Pattern DEFINITION_PATTERN = Pattern.compile(
            "^\\s*(?:def|class)\\s+([A-Za-z0-9_]+)", Pattern.MULTILINE);

    public static void main(String[] args) {
        String source = SAMPLE_CODE;
        byte[] sourceBytes = source.getBytes(java.nio.charset.StandardCharsets.UTF_8);

        int nodeCount = countApproximateNodes(source);
        System.out.println("AST node count: " + nodeCount);

        List<Definition> definitions = extractDefinitions(source, sourceBytes);
        System.out.println("\nDiscovered definitions:");
        for (Definition d : definitions) {
            System.out.printf("  [%s] %s @ line %d%n", d.kind, d.name, d.line);
        }

        String context = buildPromptContext(source, definitions);
        System.out.println("\n" + "=".repeat(50));
        System.out.println("Generated prompt context");
        System.out.println("=".repeat(50));
        System.out.println(context);
    }

    private static int countApproximateNodes(String source) {
        // Approximate AST node count by counting significant tokens.
        int count = 0;
        Matcher matcher = Pattern.compile("\\w+|[^\\w\\s]").matcher(source);
        while (matcher.find()) {
            count++;
        }
        return count;
    }

    private static List<Definition> extractDefinitions(String source, byte[] sourceBytes) {
        List<Definition> definitions = new ArrayList<>();
        Matcher matcher = DEFINITION_PATTERN.matcher(source);
        while (matcher.find()) {
            String kind = source.substring(matcher.start(1) - 4, matcher.start(1) - 1).trim();
            if ("def".equals(kind)) {
                kind = "function";
            } else if ("class".equals(kind)) {
                kind = "class";
            }
            String name = matcher.group(1);
            int line = source.substring(0, matcher.start(1)).split("\n", -1).length;
            int startByte = source.substring(0, matcher.start()).getBytes(java.nio.charset.StandardCharsets.UTF_8).length;
            int endByte = sourceBytes.length;
            definitions.add(new Definition(kind, name, line, startByte, endByte));
        }
        return definitions;
    }

    private static String buildPromptContext(String source, List<Definition> definitions) {
        List<String> lines = new ArrayList<>();
        lines.add("You are reviewing the following Python module.");
        lines.add("");
        lines.add("## Symbols");
        for (Definition d : definitions) {
            String kind = d.kind.substring(0, 1).toUpperCase() + d.kind.substring(1);
            lines.add(String.format("- %s `%s` at line %d (bytes %d-%d)",
                    kind, d.name, d.line, d.startByte, d.endByte));
        }
        lines.add("");
        lines.add("## Source");
        lines.add("```python");
        lines.add(source);
        lines.add("```");
        lines.add("");
        lines.add("Please summarize what this module does.");
        return String.join("\n", lines);
    }

    private static class Definition {
        final String kind;
        final String name;
        final int line;
        final int startByte;
        final int endByte;

        Definition(String kind, String name, int line, int startByte, int endByte) {
            this.kind = kind;
            this.name = name;
            this.line = line;
            this.startByte = startByte;
            this.endByte = endByte;
        }
    }
}

File: labs/12-lsp-treesitter/go/main.go
package main

import (
	"fmt"
	"regexp"
	"strings"
)

const sampleCode = `"""A tiny module for demonstration."""

import json
from datetime import datetime


def greet(name: str) -> str:
    """Return a friendly greeting."""
    return f"Hello, {name}!"


class Calculator:
    """Simple calculator."""

    def add(self, a: float, b: float) -> float:
        return a + b

    def subtract(self, a: float, b: float) -> float:
        return a - b


def main():
    calc = Calculator()
    print(greet("world"))
    print(calc.add(1, 2))
`

type definition struct {
	Kind      string `json:"kind"`
	Name      string `json:"name"`
	Line      int    `json:"line"`
	StartByte int    `json:"start_byte"`
	EndByte   int    `json:"end_byte"`
}

func walkTree(source string) int {
	// Placeholder for AST node count; regex extraction below gives a comparable metric.
	return len(regexp.MustCompile(`(?m)^(?:def|class)\s+`).FindAllStringIndex(source, -1))
}

func extractDefinitions(source string) []definition {
	var defs []definition
	lines := strings.Split(source, "\n")

	funcRe := regexp.MustCompile(`(?m)^def\s+([A-Za-z0-9_]+)\s*\(`)
	classRe := regexp.MustCompile(`(?m)^class\s+([A-Za-z0-9_]+)\s*[:\(]`)

	byteOffset := 0
	for i, line := range lines {
		if matches := funcRe.FindStringSubmatchIndex(line); matches != nil {
			nameStart := matches[2]
			nameEnd := matches[3]
			startByte := byteOffset + nameStart
			endByte := startByte
			// Find the end of the definition by scanning for the next top-level def/class at column 0.
			for j := i + 1; j < len(lines); j++ {
				if strings.TrimSpace(lines[j]) == "" {
					continue
				}
				if (funcRe.MatchString(lines[j]) || classRe.MatchString(lines[j])) && (len(lines[j]) == 0 || lines[j][0] != ' ') {
					endByte = byteOffset + len(line) + 1
					for k := i + 1; k < j; k++ {
						endByte += len(lines[k]) + 1
					}
					break
				}
			}
			if endByte <= startByte {
				endByte = len(source)
			}
			defs = append(defs, definition{Kind: "function", Name: line[nameStart:nameEnd], Line: i + 1, StartByte: startByte, EndByte: endByte})
		}
		if matches := classRe.FindStringSubmatchIndex(line); matches != nil {
			nameStart := matches[2]
			nameEnd := matches[3]
			startByte := byteOffset + nameStart
			endByte := startByte
			for j := i + 1; j < len(lines); j++ {
				if strings.TrimSpace(lines[j]) == "" {
					continue
				}
				if (funcRe.MatchString(lines[j]) || classRe.MatchString(lines[j])) && (len(lines[j]) == 0 || lines[j][0] != ' ') {
					endByte = byteOffset + len(line) + 1
					for k := i + 1; k < j; k++ {
						endByte += len(lines[k]) + 1
					}
					break
				}
			}
			if endByte <= startByte {
				endByte = len(source)
			}
			defs = append(defs, definition{Kind: "class", Name: line[nameStart:nameEnd], Line: i + 1, StartByte: startByte, EndByte: endByte})
		}
		byteOffset += len(line) + 1
	}

	return defs
}

func buildPromptContext(source string, defs []definition) string {
	var lines []string
	lines = append(lines, "You are reviewing the following Python module.", "", "## Symbols")
	for _, d := range defs {
		lines = append(lines, fmt.Sprintf("- %s `%s` at line %d (bytes %d-%d)", strings.Title(d.Kind), d.Name, d.Line, d.StartByte, d.EndByte))
	}
	lines = append(lines, "", "## Source", "```python", source, "```", "", "Please summarize what this module does.")
	return strings.Join(lines, "\n")
}

func main() {
	fmt.Printf("AST node count: %d\n", walkTree(sampleCode))

	defs := extractDefinitions(sampleCode)
	fmt.Println("\nDiscovered definitions:")
	for _, d := range defs {
		fmt.Printf("  [%s] %s @ line %d\n", d.Kind, d.Name, d.Line)
	}

	context := buildPromptContext(sampleCode, defs)
	fmt.Println("\n" + strings.Repeat("=", 50))
	fmt.Println("Generated prompt context")
	fmt.Println(strings.Repeat("=", 50))
	fmt.Println(context)
}
